x/rewards/types/params.go

```diff
@@ -0,0 +1,105 @@
+package types
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	paramTypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	"sigs.k8s.io/yaml"
+)
+
+var (
+	InflationRewardsRatioParamKey = []byte("InflationRewardsRatio")
+	TxFeeRebateRatioParamKey      = []byte("TxFeeRebateRatio")
+)
+
+var _ paramTypes.ParamSet = (*Params)(nil)
+
+// ParamKeyTable creates a new params table.
+func ParamKeyTable() paramTypes.KeyTable {
+	return paramTypes.NewKeyTable().RegisterParamSet(&Params{})
+}
+
+// NewParams creates a new Params instance.
+func NewParams(inflationRewardsRatio, txFeeRebateRatio sdk.Dec) Params {
+	return Params{
+		InflationRewardsRatio: inflationRewardsRatio,
+		TxFeeRebateRatio:      txFeeRebateRatio,
+	}
+}
+
+// DefaultParams returns a default set of parameters.
+func DefaultParams() Params {
+	defInflationRatio := sdk.MustNewDecFromStr("0.20")   // 20%
+	defTxFeeRebateRatio := sdk.MustNewDecFromStr("0.50") // 50%
+
+	return NewParams(defInflationRatio, defTxFeeRebateRatio)
+}
+
+// ParamSetPairs Implements the paramTypes.ParamSet interface.
+func (m *Params) ParamSetPairs() paramTypes.ParamSetPairs {
+	return paramTypes.ParamSetPairs{
+		paramTypes.NewParamSetPair(InflationRewardsRatioParamKey, &m.InflationRewardsRatio, validateInflationRewardsRatio),
+		paramTypes.NewParamSetPair(TxFeeRebateRatioParamKey, &m.TxFeeRebateRatio, validateTxFeeRebateRatio),
+	}
+}
+
+// Validate perform object fields validation.
+func (m Params) Validate() error {
+	if err := validateInflationRewardsRatio(m.InflationRewardsRatio); err != nil {
+		return err
+	}
+	if err := validateTxFeeRebateRatio(m.TxFeeRebateRatio); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// String implements the fmt.Stringer interface.
+func (m Params) String() string {
+	bz, _ := yaml.Marshal(m)
+	return string(bz)
+}
+
+func validateInflationRewardsRatio(v interface{}) (retErr error) {
+	defer func() {
+		if retErr != nil {
+			retErr = fmt.Errorf("inflationRewardsRatio param: %w", retErr)
+		}
+	}()
+
+	p, ok := v.(sdk.Dec)
+	if !ok {
+		return fmt.Errorf("invalid parameter type: %T", v)
+	}
+
+	return validateRatio(p)
+}
+
+func validateTxFeeRebateRatio(v interface{}) (retErr error) {
+	defer func() {
+		if retErr != nil {
+			retErr = fmt.Errorf("txFeeRebateRatio param: %w", retErr)
+		}
+	}()
+
+	p, ok := v.(sdk.Dec)
+	if !ok {
+		return fmt.Errorf("invalid parameter type: %T", v)
+	}
+
+	return validateRatio(p)
+}
+
+// validateRatio is a generic ratio coefficient validator.
+func validateRatio(v sdk.Dec) error {
+	if v.IsNegative() {
+		return fmt.Errorf("must be GTE 0.0")
+	}
+	if v.GTE(sdk.OneDec()) {
```

Comment:
Why one is not valid here? 

I understand it may not be valid for certain ratio, but since this is generic function, it should not say that 1 is invalid as we could add a ratio parameter in future for which 1 is valid.

Reply:
`x/mint` breaks if 0 coins are sent to FeeCollector. As for FeeRebateRatio, I'm not sure 1.0 can break anything.

Reply:
please for param validation refer to:

Let's use shorthands for the variables at play:
- GasConsumedByTX = a
- ArchInflationInBlock = b
- BlockGasLimit = c
- GasPrice = d
- GasRebateRatio = e

$${\frac{a \times b}{c} + a \times d \times e \leq a \times d}$$

Now we solve the inequality for the gas price which is `d` in our inequality.

$$
\begin{equation}
    \begin{cases}
      a > 0\\
      b >  0\\
      c > 0\\
      d \geq 0\\
      0 \leq e \leq 1\\
      \frac{a \times b}{c} + a \times d \times e \leq a \times d\\
    \end{cases}       
\end{equation}
$$

Reply:
please for param validation refer to:

Let's use shorthands for the variables at play:
- GasConsumedByTX = a
- ArchInflationInBlock = b
- BlockGasLimit = c
- GasPrice = d
- GasRebateRatio = e

$${\frac{a \times b}{c} + a \times d \times e \leq a \times d}$$

Now we solve the inequality for the gas price which is `d` in our inequality.

$$
\begin{equation}
    \begin{cases}
      a > 0\\
      b >  0\\
      c > 0\\
      d \geq 0\\
      0 \leq e \leq 1\\
      \frac{a \times b}{c} + a \times d \times e \leq a \times d\\
    \end{cases}       
\end{equation}
$$

Reply:
@fdymylja Inflation ratio can't be `1.0` because `x/mint abci` fails with `0stake: invalid coins [cosmos/cosmos-sdk@v0.45.4/x/bank/keeper/send.go:175]` it that case.

Reply:
The same for fee rebates, you can't send `0stake` from module to module. I will add those checks to our Ante handlers to prevent those panics.

Reply:
yes, I understand this part, then the redirectn logic needs to be fixed, tx fee rebates can't be 1, but tx rebate ratio can be 0, inflation can be 0 or 1, so we need to add those checks in the mintbankkeeper and the deduct fee ante to avoid sending coins if values are not valid.